ui/components/app/assets/asset-list/asset-list.tsx

```diff
@@ -104,32 +124,91 @@ const AssetList = ({ onClickAsset, showTokensLinks }: AssetListProps) => {
   // for EVM assets
   const shouldShowTokensLinks = showTokensLinks ?? isEvm;
 
-  const detectedTokensMultichain = useSelector(
-    getAllDetectedTokensForSelectedAddress,
+  const detectedTokensMultichain: {
+    [key: `0x${string}`]: Token[];
+  } = useSelector(getAllDetectedTokensForSelectedAddress);
+
+  const multichainDetectedTokensLength = Object.keys(
+    detectedTokensMultichain || {},
+  ).reduce(
+    (sum, key) => sum + detectedTokensMultichain[key as `0x${string}`].length,
+    0,
   );
 
-  const totalTokens =
-    process.env.PORTFOLIO_VIEW &&
-    !isTokenNetworkFilterEqualCurrentNetwork &&
-    detectedTokensMultichain
-      ? (Object.values(detectedTokensMultichain).reduce(
-          // @ts-expect-error TS18046: 'tokenArray' is of type 'unknown'
-          (count, tokenArray) => count + tokenArray.length,
-          0,
-        ) as number)
-      : detectedTokens.length;
+  // Add detected tokens to sate
+  useEffect(() => {
+    const importAllDetectedTokens = async () => {
+      // If autodetect tokens toggle is OFF, return
+      if (!useTokenDetection) {
+        return;
+      }
+      // TODO add event for MetaMetricsEventName.TokenAdded
+
+      if (
+        process.env.PORTFOLIO_VIEW &&
+        !isTokenNetworkFilterEqualCurrentNetwork
+      ) {
+        const importPromises = Object.entries(detectedTokensMultichain).map(
+          async ([networkId, tokens]) => {
+            const chainConfig = allNetworks[networkId as `0x${string}`];
+            const { defaultRpcEndpointIndex } = chainConfig;
+            const { networkClientId: networkInstanceId } =
+              chainConfig.rpcEndpoints[defaultRpcEndpointIndex];
+
+            await dispatch(
+              addImportedTokens(tokens as Token[], networkInstanceId),
+            );
+            tokens.forEach((importedToken) => {
+              trackEvent({
+                event: MetaMetricsEventName.TokenAdded,
+                category: MetaMetricsEventCategory.Wallet,
+                sensitiveProperties: {
+                  token_symbol: importedToken.symbol,
+                  token_contract_address: importedToken.address,
+                  token_decimal_precision: importedToken.decimals,
+                  source: MetaMetricsTokenEventSource.Detected,
+                  token_standard: TokenStandard.ERC20,
+                  asset_type: AssetType.token,
+                  token_added_type: 'detected',
+                  chain_id: chainConfig.chainId,
+                },
+              });
+            });
+          },
+        );
+
+        await Promise.all(importPromises);
+      } else {
+        await dispatch(addImportedTokens(detectedTokens, networkClientId));
```

Comment:
I think this should have a check for whether there are any detected tokens.  Otherwise it's doing an unnecessary call to the controller + state update on each load.

<img width="475" alt="Screenshot 2024-12-20 at 11 10 56 AM" src="https://github.com/user-attachments/assets/f2a3fec5-7e5e-4012-9d38-3744036e9d36" />

Reply:
maybe just changing the else to  `else if (detectedTokens?.length > 0) {`